### PR TITLE
Set `doneUserLine` when returning to the 'Program' node

### PR DESCRIPTION
**Pull request by joshlory**
Before, expressions following `ForStatement`, `WhileStatement` or `DoWhileStatement` were skipped because these nodes don't set "done" before shifting the `stateStack`:

done-step-before

After, `doneUserLine` is set when execution returns to the `Program` node:

done-step-after

This already works correctly when inside a non-global scope because we set `doneUserLine` after executing a `BlockStatement` node.

### CODE REVIEW

**Comment by joshlory** [comment]
@cpirich the main change is in ea960ddf5c40d23b9bb69e6b0133be89b68dbded (the other commit is just a refactor).

File: apps/src/JSInterpreter.js [hunk 1]
```diff
@@ -423,8 +433,10 @@ JSInterpreter.prototype.executeInterpreter = function (firstStep, runUntilCallba
     }
     var err = safeStepInterpreter(this);
     if (!err) {
-      doneUserLine = doneUserLine ||
-        (inUserCode && this.interpreter.stateStack[0] && this.interpreter.stateStack[0].done);
+      var nodeType = this.interpreter.stateStack[0].node.type;
+      if (inUserCode) {
+        doneUserLine = doneUserLine || INTERSTITIAL_NODES.hasOwnProperty(nodeType);
```
Comment (cpirich) [inline apps/src/JSInterpreter.js]: Style nit: any reason not to write `INTERSTITIAL_NODES[nodeType]`? It looks cleaner to me :smile:

**Comment by cpirich** [comment]
I like this idea. It _may_ even allow us to change the way we do unwinding/fastforwarding (we could just use a similar approach of moving forward until we find a interstitial node). For the non-maxSpeed "run" mode (not a breakpoint, simply trying to pause and paint after each line), we may not need to unwind/fastforward at all!

Some risks:
1. We used to mark `doneUserLine` much earlier. (`state.done` gets marked in a lot of different places). Of course, then we proceeded to unwind...
2. In addition, the `BlockStatement` type was considered safe to step through while unwinding. So, the net result, is that we're going to pause in a very different place. This probably matters a lot more for step-in, step-out, step-over, and breakpoints than it does for the `doneUserLine` case while running.

However, if the net result is good enough, we should be ok. I worry a little about making a bunch of changes without good test cases because if we later discover that something is broken, it's going to take a while to rewind back through all of the changes to figure out when the behavior started changing. So I do think it would be worthwhile to formalize some test cases. I'd imagine you're currently doing a lot of manual testing like I did in the early days of writing this code.

File: apps/src/JSInterpreter.js (hunk 1 above)
Comment (joshlory) [inline apps/src/JSInterpreter.js]: `hasOwnProperty` won't go up the prototype chain on misses, but I don't think we expect to have any extra junk on `Object.prototype`.  I'll change.

**Comment by joshlory** [comment]
Good catch on `state.done`, this actually regressed stopping on multiple nodes in `ForStatement` (the `update` and `test` nodes).  I'll add back the test for `state.done`.

**Comment by cpirich** [comment]
LGTM

**Pull request merged**